### PR TITLE
API to add vendor account

### PR DESCRIPTION
**Pull request by aniketrathi**
## Description
```
URI: POST /accounts/:accountId/vendor-accounts
Body: name, vendor_type: AMAZON
Response status: 201
Response Body: {
    "account": "66b0e501a9667827835aba40",
    "id": "66b1bf6dbe77e7d1618c7ce0",
    "name": "Amz-03",
    "vendor_type": "AMAZON"
}
```

## Database schema changes

```
class VendorAccount {
  account_id: str
  name: string
  vendor: VendorType
}
```

### Automated test cases added
- TestVendorAccountService
  - test_create_vendor_account
  - test_throw_exception_when_vendor_account_with_same_name_account_and_vendor_type_exist

- TestVendorAccountApi
  - test_create_vendor_account
  - test_create_vendor_account_with_same_account_name_and_vendor_type
  - test_create_vendor_account_with_same_account_and_name_but_for_different_vendor

### Manual test cases run

https://github.com/user-attachments/assets/d66f4108-100c-4fe7-a0f8-6911a5dcf461

### CODE REVIEW

File: src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from modules.vendor_account.errors import VendorAccountWithSameNameAndAccountExistsError
+from modules.vendor_account.internal.store.vendor_account_repository import VendorAccountRepository
+from modules.vendor_account.types import NameAlreadyExistParams, VendorAccount
+
+
+class VendorAccountReader:
+    @staticmethod
+    def check_name_not_exist(params: NameAlreadyExistParams) -> VendorAccount:
+        vendor_account = VendorAccountRepository.collection().find_one(
+            {"account": params.account_id, "name": params.name, "active": True}
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py]: Based on this query pattern, I think the index should be on `account_id` and `active`. In your repository, you are creating index on `account` which does not look right.

File: src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+        )
+
+        if vendor_account:
+            raise VendorAccountWithSameNameAndAccountExistsError(
+                f"Vendor account with name {params.name} already exist"
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py]: We need to write error messages that offer solution, not just the problem. Please google "How to write good error messages" and update this accordingly.

File: src/apps/backend/modules/vendor_account/internal/vendor_account_util.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from typing import Dict
+
+from modules.vendor_account.types import VendorAccount
+
+
+class VendorAccountUtil:
+    @staticmethod
+    def convert_vendor_account_db_to_vendor_account(vendor_account_db: Dict) -> VendorAccount:
+        return VendorAccount(
+            id=str(vendor_account_db["_id"]),
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_util.py:11]: order alphabetically.

File: src/apps/backend/modules/application/repository.py [hunk 4]
```diff
@@ -1,3 +1,5 @@
+import certifi
```
Comment (jjalan) [inline src/apps/backend/modules/application/repository.py]: From which module are importing this from?

File: src/apps/backend/modules/application/repository.py [hunk 5]
```diff
@@ -27,7 +29,7 @@ def get_client(cls) -> MongoClient:
   def _create_client() -> MongoClient:
     connection_uri = ConfigService.get_string('MONGODB_URI')
     Logger.info(message=f"connecting to database - {connection_uri}")
-    client = MongoClient(connection_uri, server_api=ServerApi('1'))
+    client = MongoClient(connection_uri, server_api=ServerApi('1'), tlsCAFile=certifi.where())
```
Comment (jjalan) [inline src/apps/backend/modules/application/repository.py]: Why is this change needed?

File: src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py [hunk 6]
```diff
@@ -0,0 +1,16 @@
+from modules.vendor_account.errors import VendorAccountWithSameNameAndAccountExistsError
+from modules.vendor_account.internal.store.vendor_account_repository import VendorAccountRepository
+from modules.vendor_account.types import NameAlreadyExistParams, VendorAccount
+
+
+class VendorAccountReader:
+    @staticmethod
+    def check_name_not_exist(params: NameAlreadyExistParams) -> VendorAccount:
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py]: Please move away from this pattern.

Consider exposing:

`get_venor_account_optionalaccount_id, vendor_account_name: str)`

The write method `create_vendor_account` can call this method and if this returns an instance, it can throw `VendorAccountExists` error. You can pass the `vendor_account_name` and `account` if needed and construct the message inside the error class itself.

File: src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+class VendorAccountWriter:
+    @staticmethod
+    def create_vendor_account(params: CreateVendorAccountParams):
+        VendorAccountReader.check_name_not_exist(
+            params=NameAlreadyExistParams(account_id=params.account_id, name=params.name)
+        )
+
+        vendor_account_data = {
+            "account": params.account_id,
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py]: order alphabetically here and other places.

File: src/apps/backend/modules/vendor_account/types.py
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/types.py:1]: The types.py should expose all public facing types that will be consumed by the external interface. One litmus test for this. - any typedef used int he service class, should be defined here.

File: src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+from modules.vendor_account.internal.store.vendor_account_repository import VendorAccountRepository
+from modules.vendor_account.internal.vendor_account_reader import VendorAccountReader
+from modules.vendor_account.internal.vendor_account_util import VendorAccountUtil
+from modules.vendor_account.types import CreateVendorAccountParams, NameAlreadyExistParams
+
+
+class VendorAccountWriter:
+    @staticmethod
+    def create_vendor_account(params: CreateVendorAccountParams):
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py]: define return type.

File: src/apps/backend/modules/application/repository.py (hunk 4 above)
Comment (aniketrathi) [reply src/apps/backend/modules/application/repository.py]: The `certifi` module is a Python package that provides Mozilla's CA Bundle

File: src/apps/backend/modules/application/repository.py (hunk 5 above)
Comment (aniketrathi) [reply src/apps/backend/modules/application/repository.py]: I was facing the Mongo SSL certificate issue. So that's why I added this change. For more reference please see [this](https://www.mongodb.com/community/forums/t/serverselectiontimeouterror-ssl-certificate-verify-failed-trying-to-understand-the-origin-of-the-problem/115288).

Comment (jjalan) [reply src/apps/backend/modules/application/repository.py]: We should understand why other products, engineers or platform has not baked this into the boilerplate. Can you please start a thread with platform team to understand this better, and either (a) Get similar change merged in boilerplate (b) Remove it from here.

Comment (aniketrathi) [reply src/apps/backend/modules/application/repository.py]: sure

File: src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py (hunk 1 above)
Comment (aniketrathi) [reply src/apps/backend/modules/vendor_account/internal/vendor_account_reader.py]: I think it should be on `account` , `name` and `active`.

File: src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+from modules.vendor_account.types import CreateVendorAccountParams
+from modules.vendor_account.types import VendorAccount
+from modules.vendor_account.errors import VendorAccountWithSameNameAndAccountExistsError
+
+
+class VendorAccountWriter:
+    @staticmethod
+    def create_vendor_account(params: CreateVendorAccountParams) -> VendorAccount:
+        vendor_account = VendorAccountReader.get_vendor_account_optional(
+            account_id=params.account_id, vendor_account_name=params.name
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py]: I believe you also want to scope the search by `vendor_type`. Maybe Ok to have same name across different vendors. You might have to update the index and add new tests around this as well

File: src/apps/backend/modules/vendor_account/types.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+from modules.account.types import Account
+
+
+class VendorType(Enum):
+    AMAZON = "AMAZON"
+
+
+@dataclass(frozen=True)
+class VendorAccount:
+    account: str | Account
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/types.py]: This is not right.

It should be `account_id`: str and `account: Account` if we must include full `account` object.

Why do we need it through?

File: src/apps/backend/modules/vendor_account/internal/store/vendor_account_repository.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+from pymongo import ASCENDING
+from modules.application.repository import ApplicationRepository
+
+
+class VendorAccountRepository(ApplicationRepository):
+    collection_name = "vendor_accounts"
+
+    @classmethod
+    def on_init_collection(cls, collection):
+        collection.create_index([("account", ASCENDING), ("active", ASCENDING), ("name", ASCENDING)])
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/store/vendor_account_repository.py]: `account_id`

File: src/apps/backend/modules/vendor_account/internal/vendor_account_util.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
+from typing import Dict
+
+from modules.vendor_account.types import VendorAccount
+
+
+class VendorAccountUtil:
+    @staticmethod
+    def convert_vendor_account_db_to_vendor_account(vendor_account_db: Dict) -> VendorAccount:
+        return VendorAccount(
+            account=str(vendor_account_db["account"]),
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_util.py]: account_id

File: src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        )
+
+        if vendor_account:
+            raise VendorAccountWithSameNameAndAccountExistsError(vendor_account_name=params.name)
+
+        vendor_account_data = {
+            "account": ObjectId(params.account_id),
```
Comment (jjalan) [inline src/apps/backend/modules/vendor_account/internal/vendor_account_writer.py]: account_id

**Review by jjalan — APPROVED** [review]
Changes look good.

Before merging, please make sure:

(a) PR title is updated
(b) PR description has demo video of postman
(c) CI pipeline is passing.

**Comment by aniketrathi** [comment]
> Changes look good.
> 
> Before merging, please make sure:
> 
> (a) PR title is updated (b) PR description has demo video of postman (c) CI pipeline is passing.

We can't run CI pipeline here for now.
